internal/statemachine/classic_states.go

```diff
@@ -583,21 +604,66 @@ func (stateMachine *StateMachine) verifyArtifactNames() error {
 	stateMachine.VolumeNames = make(map[string]string)
 
 	if len(stateMachine.GadgetInfo.Volumes) > 1 {
-		// dereferencing is safe because this state only runs when Img is non-nil
-		for _, img := range *classicStateMachine.ImageDef.Artifacts.Img {
-			if img.ImgVolume == "" {
-				return fmt.Errorf("Volume names must be specified for each image when using a gadget with more than one volume")
+		// first handle .img files if they are specified
+		if classicStateMachine.ImageDef.Artifacts.Img != nil {
+			for _, img := range *classicStateMachine.ImageDef.Artifacts.Img {
+				if img.ImgVolume == "" {
+					return fmt.Errorf("Volume names must be specified for each image when using a gadget with more than one volume")
+				}
+				stateMachine.VolumeNames[img.ImgVolume] = img.ImgName
+			}
+		}
+		if classicStateMachine.ImageDef.Artifacts.Qcow2 != nil {
+			for _, qcow2 := range *classicStateMachine.ImageDef.Artifacts.Qcow2 {
+				if qcow2.Qcow2Volume == "" {
+					return fmt.Errorf("Volume names must be specified for each image when using a gadget with more than one volume")
+				}
+				// We can save a whole lot of disk I/O here if the volume is already specified as a .img file
+				if classicStateMachine.ImageDef.Artifacts.Img != nil {
+					found := false
+					for _, img := range *classicStateMachine.ImageDef.Artifacts.Img {
+						if img.ImgVolume == qcow2.Qcow2Volume {
+							found = true
+						}
+					}
+					if !found {
+						stateMachine.VolumeNames[qcow2.Qcow2Volume] = fmt.Sprintf("%s.img", qcow2.Qcow2Name)
```

Comment:
Now I understand what you meant about the logic for detecting existing raw images complicating the code slightly. It's not bad though, but I know I had to sit down for a minute or two to wrap my head around this bit here - since it might look confusing that we're dealing with qcow2 images but the name put into `VolumeNames` is a `.img` filename. I think it might be worthwhile to mention here in the comments that in cases of qcow2 artifacts, the VolumeNames field is used as 'input files' for the `qemu-img` command rather than 'output filenames' as one would expect normally.

But we'll have to revisit this at some point, since I'm worried the logic will get even more complicated if we add more output formats that base on converting the raw image.

Reply:
yeah if we add any more I think we'll need a helper function to handle this logic.